core/src/main/java/org/infinispan/persistence/support/NonBlockingStoreAdapter.java

```diff
@@ -0,0 +1,340 @@
+            characteristics.add(Characteristic.WRITE_ONLY);
+         }
+
+         if (storeImpl instanceof AdvancedCacheWriter) {
+            characteristics.add(Characteristic.EXPIRATION);
+         } else if (!(storeImpl instanceof CacheWriter)) {
+            characteristics.add(Characteristic.READ_ONLY);
+         }
+      }
+
+      Store storeAnnotation = storeImpl.getClass().getAnnotation(Store.class);
+      if (storeAnnotation != null && storeAnnotation.shared()) {
+         characteristics.add(Characteristic.SHAREABLE);
+      }
+
+      // Transactional is a special interface that could be true on a segment or not segmented store both
+      if (storeImpl instanceof TransactionalCacheWriter) {
+         characteristics.add(Characteristic.TRANSACTIONAL);
+      }
+      return characteristics;
+   }
+
+
+
+   @Override
+   public CompletionStage<Void> start(InitializationContext ctx) {
+      blockingManager = ctx.getBlockingManager();
+      marshallableEntryFactory = ctx.getMarshallableEntryFactory();
+      return blockingManager.runBlocking(() -> {
+         if (isReadOnly()) {
+            loader().init(ctx);
+         } else {
+            writer().init(ctx);
+         }
+         oldStoreImpl.start();
+      }, nextTraceId());
+   }
+
+   @Override
+   public CompletionStage<Void> stop() {
+      return blockingManager.runBlocking(oldStoreImpl::stop, nextTraceId());
+   }
+
+   @Override
+   public Set<Characteristic> characteristics() {
+      return characteristics;
+   }
+
+   @Override
+   public MediaType getKeyMediaType(MediaType storageMediaType, Set<MediaType> supportedMediaTypes) {
+      return MediaType.MATCH_ALL;
+   }
+
+   @Override
+   public MediaType getValueMediaType(MediaType storageMediaType, Set<MediaType> supportedMediaTypes) {
+      return MediaType.MATCH_ALL;
+   }
+
+   @Override
+   public CompletionStage<Long> size(IntSet segments) {
+      return blockingManager.supplyBlocking(() ->
+            isSegmented() ? segmentedStore().size(segments) : advancedLoader().size(), nextTraceId())
+            .thenApply(Integer::longValue);
+   }
+
+   @Override
+   public CompletionStage<Long> approximateSize(IntSet segments) {
+      // Old SPI didn't support approximations
+      return size(segments);
+   }
+
+   @Override
+   public Publisher<MarshallableEntry<K, V>> publishEntries(IntSet segments, Predicate<? super K> filter, boolean includeValues) {
+      Publisher<MarshallableEntry<K, V>> publisher;
+      if (isSegmented()) {
+         publisher = segmentedStore().entryPublisher(segments, filter, includeValues, true);
+      } else {
+         publisher = advancedLoader().entryPublisher(filter, includeValues, true);
+      }
+      // Despite this being a publisher, we assume the subscription is blocking as the SPI never enforced this
+      // We do however assume the creation of the Publisher is not blocking... maybe we should?
+      return blockingManager.blockingPublisher(publisher);
```

Comment:
I would say we probably should err on the side of caution and assume that the creation blocks.

Reply:
Yeah I was having a hard time with this one, because it is a lot of context switching to create it on the blocking thread and then subscribe again on the blocking thread. All of our impls do not block for the creation, just the subscription.

But I guess a user may not know how to implement Publisher correctly. For example they may acquire the DB connection in this method instead of on subscription.

But then again I don't think a user will be implementing the old SPI moving forward, hopefully. I plan on adding a bunch of examples of how to implement the SPI given specific use cases for Final.

Reply:
I have not changed this btw. I think it is safe enough to assume it isn't blocking for now. And as mentioned this is only for the old SPI.